File: src/main/java/com/kisman/cc/features/module/render/shader/Shader.java
package com.kisman.cc.features.module.render.shader;

import org.apache.commons.io.IOUtils;
import org.lwjgl.opengl.*;

import java.io.InputStream;
import java.nio.charset.Charset;
import java.util.*;

public abstract class Shader {
    public String name;
    public int program;
    public Map<String, Integer> uniformsMap;

    public Shader(String fragmentShader) {
        this.name = fragmentShader;

        int vertexShaderID = 0;
        int fragmentShaderID = 0;

        try {
            InputStream vertexStream = this.getClass().getResourceAsStream("/assets/kismancc/shader/vertex.vert");
            if(vertexStream != null) vertexShaderID = this.createShader(IOUtils.toString(vertexStream, Charset.defaultCharset()), ARBVertexShader.GL_VERTEX_SHADER_ARB);
            IOUtils.closeQuietly(vertexStream);
            InputStream fragmentStream = this.getClass().getResourceAsStream("/assets/kismancc/shader/fragment/" + fragmentShader);
            if(fragmentStream != null) fragmentShaderID = this.createShader(IOUtils.toString(fragmentStream, Charset.defaultCharset()), ARBFragmentShader.GL_FRAGMENT_SHADER_ARB);
            IOUtils.closeQuietly(fragmentStream);
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        if (vertexShaderID == 0 || fragmentShaderID == 0) return;

        program = ARBShaderObjects.glCreateProgramObjectARB();
        if (program == 0) return;

        ARBShaderObjects.glAttachObjectARB(program, vertexShaderID);
        ARBShaderObjects.glAttachObjectARB(program, fragmentShaderID);
        ARBShaderObjects.glLinkProgramARB(program);
        ARBShaderObjects.glValidateProgramARB(program);
    }

    public void startShader() {
        GL11.glPushMatrix();
        GL20.glUseProgram(program);
        if (uniformsMap == null) {
            uniformsMap = new HashMap<>();
            setupUniforms();
        }
        updateUniforms();
    }

    public void stopShader() {
        GL20.glUseProgram(0);
        GL11.glPopMatrix();
    }

    public void setupUniforms() {}
    public void updateUniforms() {}

    public int createShader(String shaderSource, int shaderType) {
        int shader = 0;
        try {
            shader = ARBShaderObjects.glCreateShaderObjectARB(shaderType);
            if (shader == 0) return 0;
            ARBShaderObjects.glShaderSourceARB(shader, shaderSource);
            ARBShaderObjects.glCompileShaderARB(shader);
            if (ARBShaderObjects.glGetObjectParameteriARB(shader, ARBShaderObjects.GL_OBJECT_COMPILE_STATUS_ARB) == GL11.GL_FALSE) throw new RuntimeException("Error creating shader: " + this.getLogInfo(shader));
            return shader;
        } catch (Exception e) {
            ARBShaderObjects.glDeleteObjectARB(shader);
            throw e;
        }
    }

    public String getLogInfo(int i) {
        return ARBShaderObjects.glGetInfoLogARB(i, ARBShaderObjects.glGetObjectParameteriARB(i, ARBShaderObjects.GL_OBJECT_INFO_LOG_LENGTH_ARB));
    }

    public void setUniform(String uniformName, int location) {
        uniformsMap.put(uniformName, location);
    }

    public void setupUniform(String uniformName) {
        setUniform(uniformName, GL20.glGetUniformLocation(this.program, uniformName));
    }

    public int getUniform(String uniformName) {
        return uniformsMap.get(uniformName);
    }

    public int getProgramId() {
        return program;
    }
}

File: src/main/java/com/kisman/cc/util/render/shader/framebuffer/Shader.kt
package com.kisman.cc.util.render.shader.framebuffer

import org.apache.commons.io.IOUtils
import org.lwjgl.opengl.ARBShaderObjects
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL20
import java.nio.charset.Charset

/**
 * @author _kisman_
 * @since 20:36 of 17.08.2022
 */
abstract class Shader(
    fragmentShader : String
) {
    val uniformsRaw = HashMap<String, Int>()
    var program = 0


    init {
        run {
            var vertexShaderID = 0
            var fragmentShaderID = 0

            try {
                /*final InputStream vertexStream = this.getClass().getResourceAsStream("/assets/kismancc/shader/vertex.vert");
            vertexShaderID = this.createShader(IOUtils.toString(vertexStream, Charset.defaultCharset()), 35633);
            IOUtils.closeQuietly(vertexStream);
            final InputStream fragmentStream = this.getClass().getResourceAsStream("/assets/kismancc/shader/fragment/" + fragmentShader);
            fragmentShaderID = this.createShader(IOUtils.toString(fragmentStream, Charset.defaultCharset()), 35632);
            IOUtils.closeQuietly(fragmentStream);*/
                val vertexStream = javaClass.getResourceAsStream(vertex())
                vertexShaderID = createShader(IOUtils.toString(vertexStream, Charset.defaultCharset()), 35633)
                IOUtils.closeQuietly(vertexStream)

                val fragmentStream = javaClass.getResourceAsStream(fragment(fragmentShader))
                fragmentShaderID = createShader(IOUtils.toString(fragmentStream, Charset.defaultCharset()), 35632)
                IOUtils.closeQuietly(fragmentStream)
            } catch(e : Exception) {
                e.printStackTrace()

                return@run
            }

            if(vertexShaderID == 0 || fragmentShaderID == 0) {
                return@run
            }

            program = ARBShaderObjects.glCreateProgramObjectARB()

            if(program == 0) {
                return@run
            }

            ARBShaderObjects.glAttachObjectARB(
                program,
                vertexShaderID
            )

            ARBShaderObjects.glAttachObjectARB(
                program,
                fragmentShaderID
            )

            ARBShaderObjects.glLinkProgramARB(program)
            ARBShaderObjects.glValidateProgramARB(program)
        }
    }

    fun startShader() {
        GL11.glPushMatrix()
        GL20.glUseProgram(program)

        uniformsRaw.clear()

        setupUniforms()
        updateUniforms()
    }

    fun stopShader() {
        GL20.glUseProgram(0)
        GL11.glPopMatrix()
    }

    abstract fun setupUniforms()
    abstract fun updateUniforms()

    private fun createShader(
        source : String,
        type : Int
    ) : Int {
        var shader = 0

        try {
            shader = ARBShaderObjects.glCreateShaderObjectARB(type)

            if(shader == 0) {
                return 0
            }

            ARBShaderObjects.glShaderSourceARB(
                shader,
                source
            )

            ARBShaderObjects.glCompileShaderARB(shader)

            if(ARBShaderObjects.glGetObjectParameteriARB(shader, 35713) == 0) {
                throw RuntimeException(
                    "Error creating shader: ${getLogInfo(shader)}"
                )
            }

            return shader
        } catch(e : Exception) {
            ARBShaderObjects.glDeleteObjectARB(shader)

            throw e
        }
    }

    private fun getLogInfo(i : Int) : String = ARBShaderObjects.glGetInfoLogARB(i, ARBShaderObjects.glGetObjectParameteriARB(i, 35716))

    fun setupUniform(
        name : String
    ) {
        uniformsRaw[name] = GL20.glGetUniformLocation(
            program,
            name
        )
    }
}